File: recyclerview/src/main/java/com/kfaraj/support/util/SparseLongArray.java
package com.kfaraj.support.util;

import android.os.Parcel;
import android.os.Parcelable;
import android.util.SparseArray;

/**
 * Maps integers to longs.
 */
public class SparseLongArray extends SparseArray<Long> implements Parcelable {

    /**
     * The creator.
     */
    public static final Creator<SparseLongArray> CREATOR = new Creator<SparseLongArray>() {
        @Override
        public SparseLongArray createFromParcel(Parcel source) {
            return new SparseLongArray(source);
        }
        @Override
        public SparseLongArray[] newArray(int size) {
            return new SparseLongArray[size];
        }
    };

    /**
     * Constructor
     */
    public SparseLongArray() {
    }

    /**
     * Constructor
     *
     * @param source the source.
     */
    private SparseLongArray(Parcel source) {
        final int size = source.readInt();
        for (int i = 0; i < size; i++) {
            final int key = source.readInt();
            final Long value = (Long) source.readValue(Long.class.getClassLoader());
            put(key, value);
        }
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public int describeContents() {
        return 0;
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public void writeToParcel(Parcel dest, int flags) {
        final int size = size();
        dest.writeInt(size);
        for (int i = 0; i < size; i++) {
            final int key = keyAt(i);
            final Long value = valueAt(i);
            dest.writeInt(key);
            dest.writeValue(value);
        }
    }

}

File: recyclerview/src/main/java/com/kfaraj/support/util/SparseLongArray.kt
package com.kfaraj.support.util

import android.os.Parcel
import android.os.Parcelable
import android.util.SparseArray

/**
 * Maps integers to longs.
 */
internal class SparseLongArray : SparseArray<Long?>,
    Parcelable {

    /**
     * Constructor.
     */
    constructor()

    /**
     * Constructor.
     *
     * @param source the source.
     */
    private constructor(source: Parcel) {
        val size = source.readInt()
        for (i in 0..<size) {
            val key = source.readInt()
            val value = source.readValue(Long::class.java.classLoader) as Long?
            put(key, value)
        }
    }

    override fun describeContents(): Int {
        return 0
    }

    override fun writeToParcel(dest: Parcel, flags: Int) {
        val size = size()
        dest.writeInt(size)
        for (i in 0..<size) {
            val key = keyAt(i)
            val value = valueAt(i)
            dest.writeInt(key)
            dest.writeValue(value)
        }
    }

    /**
     * The creator.
     */
    companion object CREATOR : Parcelable.Creator<SparseLongArray> {
        override fun createFromParcel(source: Parcel): SparseLongArray {
            return SparseLongArray(source)
        }

        override fun newArray(size: Int): Array<SparseLongArray?> {
            return arrayOfNulls(size)
        }
    }

}
